JDBI 3 - Handle.close() throws Failed to clear transaction status on close error

We use jdbi api in our drop wizard based application which internally uses tomcat connection pool for db operations.
Below code snippet is used to delete user from our oracle database.

```
jdbi.useHandle(
 handle ->
 handle.createUpdate("DROP USER XXXXX CASCADE")
 .execute());
```

The above code throws an error -

```
Caused by: org.jdbi.v3.core.CloseException: Failed to clear transaction status on close
at org.jdbi.v3.core.Handle.close(Handle.java:183)
at org.jdbi.v3.core.Jdbi.withHandle(Jdbi.java:358)
at org.jdbi.v3.core.Jdbi.useHandle(Jdbi.java:373)
at com.test.db.DBSchemaProviderImpl.deleteDBSchema(DBSchemaProviderImpl.java:57)
... 7 common frames omitted
Caused by: org.jdbi.v3.core.transaction.TransactionException: Failed to test for transaction status
at org.jdbi.v3.core.transaction.LocalTransactionHandler$BoundLocalTransactionHandler.isInTransaction(LocalTransactionHandler.java:194)
at org.jdbi.v3.core.Handle.isInTransaction(Handle.java:346)
at org.jdbi.v3.core.Handle.close(Handle.java:154)
... 10 common frames omitted
Caused by: java.sql.SQLException: Connection has already been closed.
at org.apache.tomcat.jdbc.pool.ProxyConnection.invoke(ProxyConnection.java:121)
at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:109)
at org.apache.tomcat.jdbc.pool.interceptor.AbstractCreateStatementInterceptor.invoke(AbstractCreateStatementInterceptor.java:79)
at org.apache.tomcat.jdbc.pool.JdbcInterceptor.invoke(JdbcInterceptor.java:109)
at org.apache.tomcat.jdbc.pool.DisposableConnectionFacade.invoke(DisposableConnectionFacade.java:81)
at com.sun.proxy.$Proxy110.getAutoCommit(Unknown Source)
at org.jdbi.v3.core.transaction.LocalTransactionHandler$BoundLocalTransactionHandler.isInTransaction(LocalTransactionHandler.java:192)
... 12 common frames omitted
```


Please suggest a way of working with this code. Thanks

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -1,5 +1,7 @@
 # Unreleased
 
+  - Fix exception in Handle#close() when underlying connection is closed (#2065)
+
 # 3.31.0
 
   - Support binding parameters of type CharSequence (#2057, thanks @sman-81)

--- a/core/src/main/java/org/jdbi/v3/core/Handle.java
+++ b/core/src/main/java/org/jdbi/v3/core/Handle.java
@@ -148,14 +148,14 @@ public class Handle implements Closeable, Configurable<Handle> {
             return;
         }
 
-        boolean connectionIsLive = false;
+        boolean connectionIsLive;
 
         try {
             connectionIsLive = !connection.isClosed();
         } catch (SQLException e) {
             // if the connection state can not be determined, assume that the
-            // connection is closed
-            suppressed.add(e);
+            // connection is closed and ignore the exception
+            connectionIsLive = false;
         }
 
         boolean wasInTransaction = false;

--- a/core/src/main/java/org/jdbi/v3/core/Handle.java
+++ b/core/src/main/java/org/jdbi/v3/core/Handle.java
@@ -148,8 +148,19 @@ public class Handle implements Closeable, Configurable<Handle> {
             return;
         }
 
+        boolean connectionIsLive = false;
+
+        try {
+            connectionIsLive = !connection.isClosed();
+        } catch (SQLException e) {
+            // if the connection state can not be determined, assume that the
+            // connection is closed
+            suppressed.add(e);
+        }
+
         boolean wasInTransaction = false;
-        if (forceEndTransactions && localConfig.get().get(Handles.class).isForceEndTransactions()) {
+
+        if (connectionIsLive && forceEndTransactions && localConfig.get().get(Handles.class).isForceEndTransactions()) {
             try {
                 wasInTransaction = isInTransaction();
             } catch (Exception e) {
@@ -168,14 +179,18 @@ public class Handle implements Closeable, Configurable<Handle> {
             }
         }
 
-        try {
-            statementBuilder.close(getConnection());
-        } catch (Exception e) {
-            suppressed.add(e);
+        if (connectionIsLive) {
+            try {
+                statementBuilder.close(getConnection());
+            } catch (Exception e) {
+                suppressed.add(e);
+            }
         }
 
         try {
-            closer.close(connection);
+            if (connectionIsLive) {
+                closer.close(connection);
+            }
 
             if (!suppressed.isEmpty()) {
                 final Throwable original = suppressed.remove(0);

--- a/core/src/test/java/org/jdbi/v3/core/TestHandle.java
+++ b/core/src/test/java/org/jdbi/v3/core/TestHandle.java
@@ -175,6 +175,22 @@ public class TestHandle {
         }
     }
 
+    @Test
+    public void testIssue2065() throws Exception {
+        Handle h = h2Extension.openHandle();
+
+        h.begin();
+        h.execute("insert into something (id, name) values (1, 'Brian')");
+        String value = h.createQuery("select name from something where id = 1").mapToBean(Something.class).one().getName();
+        h.commit();
+
+        // close connection between commit and handle close - this may be done by the connection pool or timeout or sth else
+        h.getConnection().close();
+
+        h.close();
+        assertThat(value).isEqualTo("Brian");
+    }
+
     static class BoomHandler extends LocalTransactionHandler {
         boolean failTest;
         boolean failRollback;

